### PR TITLE
ZCS-2344 unsubscribe shared

### PR DESCRIPTION
**Pull request by fatkudu**

### CODE REVIEW

**Comment by fatkudu** [comment]
Made a couple more small changes.  Tests were failing when I rebased - this seemed to be down to SaveIMAPSubscriptionsRequest - not sure why they weren't failing before I left.
Was also seeing NPEs in the logs, so tidied up ImapRemoteSession code to avoid that and give context as to what would have caused an NPE.

File: store/src/java/com/zimbra/cs/imap/ImapCredentials.java [hunk 1]
```diff
@@ -159,7 +160,8 @@ private void saveSubscriptions(Set<String> subscriptions) throws ServiceExceptio
     }
 
     protected Set<String> listSubscriptions() throws ServiceException {
-        return getImapMailboxStore().listSubscriptions(getContext());
+        Set<String> subs = getImapMailboxStore().listSubscriptions(getContext());
```
Comment (grishick) [inline store/src/java/com/zimbra/cs/imap/ImapCredentials.java]: not sure why this this change is needed

Comment (fatkudu) [reply store/src/java/com/zimbra/cs/imap/ImapCredentials.java]: I'll add comments for clarity and squash the commit with the previous change.  This is actually the cause of the user visible problem addressed by the bug - was later trying to modify the returned set when subscribing/unsubscribing and that doesn't go too well when the set is not modifiable.

**Comment by fatkudu** [comment]
@grishick regarding your comment:
> please file another bug about "/" prefix. I think it's not breaking clients, but it's annoying that we have to work around it in tests.

I've researched this a bit more and decided it is a misunderstanding on my part.  For most IMAP mailboxes, there is no '/' at the beginning of the name - this differs from the paths on our Folder objects but that is OK.  For IMAP mailboxes in the Other Users' Namespace (i.e. start with "/home/") they do start with '/' but that is because we use the prefix "/home/" and it comes from the start of that prefix.  I'm updating the test code to reflect this and will submit those changes soon.

File: store/src/java/com/zimbra/qa/unittest/ImapTestBase.java [hunk 2]
```diff
@@ -326,16 +343,38 @@ protected void doListShouldFail(ImapConnection conn, String ref, String mailbox,
...
-                    ref, mailbox), listResult);
-            assertEquals(String.format( "Number of entries in list returned for 'list \"%s\" \"%s\"'",
-                ref, mailbox), expected, listResult.size());
+            assertNotNull(String.format("list result %s should not be null", cmdDesc), listResult);
+            assertEquals(String.format( "Number of entries in list returned for %s", cmdDesc),
+                    expected, listResult.size());
+            return listResult;
+        } catch (CommandFailedException cfe) {
+            String err = cfe.getError();
+            fail(String.format("cmdDesc returned error '%s'", cmdDesc, err));
```
[the remark this answers is not in the document]
Comment (fatkudu) [reply store/src/java/com/zimbra/qa/unittest/ImapTestBase.java]: Good catch!  Thanks.

[the remark this answers is not in the document]
Comment (fatkudu) [reply store/src/java/com/zimbra/qa/unittest/ImapTestBase.java]: Should be fixed now

**Pull request merged**